Java build broken?

*The default build for harviewer, `ant build`, uses Node.JS.  This is the preferred way to build harviewer.  But the following issue suggests that the Java-based build is now broken.*

-----

Just tried `ant -Drjs.runner=java build` and it hangs for me.  I waited about 20 seconds and it was stuck here:

     [java] Tracing dependencies for: harViewer
     [java]
     [java] Tracing dependencies for: harPreview
     [java] Uglify file: D:/dev/git_repos/harviewer2/webapp-build/har.js

Running [jvisualvm](http://docs.oracle.com/javase/6/docs/technotes/guides/visualvm/applications_local.html) I could attach to the `org.mozilla.javascript.tools.shell.Main` code and saw this:

![harviewer-57-jvisualvm](https://cloud.githubusercontent.com/assets/1020541/15686411/0fc1a06c-2768-11e6-96c6-e1c59455f1a1.png)

so a lot of time is being burnt up in the [r.js](https://github.com/requirejs/r.js) code (top row in the Hot Spots).

The CPU is 100% (for a single core) and the build seems to have entered an infinite loop.

-----

- edit - Is the build hanging or running slowly? - https://github.com/mishoo/UglifyJS2/pull/1024

@gitgrimbo, I failed at running: ant -Drjs.runner=java build due to I/O Error as below:
D:\Software\test_tool\Web\harviewer-master\harviewer-master>ant -Drjs.runner=java build
Buildfile: D:\Software\test_tool\Web\harviewer-master\harviewer-master\build.xml

check-rjs-folder:

get-rjs:

rjs-determine-runner:

rjs-java:
     [java] js: "D:\Software\test_tool\Web\harviewer-master\harviewer-master/r.js/dist/r.js", line 1
3991: unterminated string literal
     [java] js:         expression: "[AST_Node] the `switch` 鈥渄iscriminant鈥?
     [java] js: ...........................................................^
     [java] Java Result: 3

rjs-node:

build:
     [echo] Building HAR Viewer on:
     [echo] os.name = Windows 7
     [echo] os.arch = amd64
     [echo] os.version = 6.1
     [copy] Copying 1 file to D:\Software\test_tool\Web\harviewer-master\harviewer-master\webapp-bui
ld\scripts\preview
     [copy] Copying 1 file to D:\Software\test_tool\Web\harviewer-master\harviewer-master\webapp-bui
ld\scripts\domplate
    [jsmin] Minimizing D:\Software\test_tool\Web\harviewer-master\harviewer-master\webapp-build\scri
pts\domplate\domplate.js

BUILD FAILED
D:\Software\test_tool\Web\harviewer-master\harviewer-master\build.xml:186: I/O Error when preprocess
ing file

Total time: 3 seconds

--- a/README.md
+++ b/README.md
@@ -25,6 +25,7 @@ Components
 
 Build Tools
 -----------
+* Java 8 (for [Nashorn/jjs](http://docs.oracle.com/javase/8/docs/technotes/tools/windows/jjs.html) build)
 * Ant: http://ant.apache.org/
 * js-build-tools: http://code.google.com/p/js-build-tools/
 * js-min (ant task): http://code.google.com/p/jsmin-ant-task/
@@ -40,12 +41,35 @@ Testing
 
 Development
 -----------
+
+A build can be performed in one of three ways.
+
+### 1) Node.js
+
+This is the default and preferred method.  Node.js must be installed.
 To start a build, go to the project directory and execute the following command:
 
 `ant build`
 
-This default build requires Node.js to be installed (the preferred option).
+(*Indicative build time - about 9 seconds.  Node.js v6.0.0*)
 
-For a Java-only build (not the preferred option), execute the following command:
+### 2) Java 8 / Nashorn
+
+To build using the [jjs](https://docs.oracle.com/javase/8/docs/technotes/tools/windows/jjs.html) command, Java 8 must be installed.
+Execute the following command:
+
+`ant -Drjs.runner=nashorn build`
+
+(*Indicative build time - about 1 minute 37 seconds.  nashorn 1.8.0_92*)
+
+### 3) Pre-Java 8 / Rhino
+
+To build using Java 7 or earlier, execute the following command:
 
 `ant -Drjs.runner=java build`
+
+You will have to set the [optimize build configuration value](https://github.com/janodvarko/harviewer/blob/0997957b3ecb9fbdb27df4260d5bc85c653fac81/webapp/scripts/app.build.js#L11) to `"closure"`, for example:
+
+     optimize: "closure",
+
+(*Indicative build time - about 42 seconds*)

--- a/build.xml
+++ b/build.xml
@@ -117,15 +117,20 @@
             <matches pattern="^node$" string="${rjs.runner}"/>
         </condition>
 
+        <condition property="rjs.runner.isNashorn">
+            <matches pattern="^nashorn$" string="${rjs.runner}"/>
+        </condition>
+
         <condition property="rjs.runner.isJava">
             <matches pattern="^java$" string="${rjs.runner}"/>
         </condition>
 
-        <fail message="Ant property 'rjs.runner' must be set to either 'node' (the default) or 'java'. Value was '${rjs.runner}'.">
+        <fail message="Ant property 'rjs.runner' must be set to either 'node' (the default), 'nashorn' or 'java'. Value was '${rjs.runner}'.">
             <condition>
                 <not>
                     <or>
                         <isset property="rjs.runner.isNode"/>
+                        <isset property="rjs.runner.isNashorn"/>
                         <isset property="rjs.runner.isJava"/>
                     </or>
                 </not>
@@ -141,6 +146,16 @@
     <target name="rjs-node" depends="rjs-determine-runner" if="rjs.runner.isNode">
         <exec dir="${app.dir}/scripts" executable="node" resolveexecutable="true">
             <arg value="${rjs.cmd}"/>
+            <arg value="-o"/>
+            <arg value="app.build.js"/>
+        </exec>
+    </target>
+
+    <target name="rjs-nashorn" depends="rjs-determine-runner" if="rjs.runner.isNashorn">
+        <exec dir="${app.dir}/scripts" executable="jjs" resolveexecutable="true">
+            <arg value="-scripting"/>
+            <arg value="${rjs.cmd}"/>
+            <arg value="--"/>
             <arg value="-o"/>
             <arg value="app.build.js"/>
         </exec>
@@ -163,7 +178,7 @@
 
 
     <!-- Build HAR Viewer package (the result is within build.dir) -->
-    <target name="build" depends="rjs-java, rjs-node">
+    <target name="build" depends="rjs-java, rjs-node, rjs-nashorn">
 
         <!-- Log info about the current OS -->
         <echo message="Building HAR Viewer on:" />

